libr/util/json.c

```diff
@@ -0,0 +1,374 @@
+			p++;
+			break;
+		case '{':
+			js = create_json (R_JSON_OBJECT, key, parent);
+			p++;
+			while (1) {
+				const char *new_key;
+				p = parse_key (&new_key, p);
+				if (!p) {
+					return 0; // error
+				}
+				if (*p == '}') {
+					return p + 1; // end of object
+				}
+				p = parse_value (js, new_key, p);
+				if (!p) {
+					return 0; // error
+				}
+			}
+		case '[':
+			js = create_json (R_JSON_ARRAY, key, parent);
+			p++;
+			while (1) {
+				p = parse_value (js, 0, p);
+				if (!p) {
+					return 0; // error
+				}
+				if (*p == ']') {
+					return p + 1; // end of array
+				}
+			}
+		case ']':
+			return p;
+		case '"':
+			p++;
+			js = create_json (R_JSON_STRING, key, parent);
+			js->str_value = unescape_string (p, &p);
+			if (!js->str_value) {
+				return 0; // propagate error
+			}
+			return p;
+		case '-':
+		case '0':
+		case '1':
+		case '2':
+		case '3':
+		case '4':
+		case '5':
+		case '6':
+		case '7':
+		case '8':
+		case '9': {
+			js = create_json (R_JSON_INTEGER, key, parent);
+			errno = 0;
+			char *pe;
+			if (*p == '-') {
+				js->num.s_value = (st64)strtoll (p, &pe, 0);
+			} else {
+				js->num.u_value = (ut64)strtoull (p, &pe, 0);
+			}
+			if (pe == p || errno == ERANGE) {
+				R_JSON_REPORT_ERROR ("invalid number", p);
+				return 0; // error
+			}
+			if (*pe == '.' || *pe == 'e' || *pe == 'E') { // double value
+				js->type = R_JSON_DOUBLE;
+				errno = 0;
+				js->num.dbl_value = strtod (p, &pe);
+				if (pe == p || errno == ERANGE) {
+					R_JSON_REPORT_ERROR ("invalid fractional number", p);
+					return 0; // error
+				}
+			} else {
+				if (*p == '-') {
+					js->num.dbl_value = js->num.s_value;
+				} else {
+					js->num.dbl_value = js->num.u_value;
+				}
+			}
+			return pe;
+		}
+		case 't':
+			if (!strncmp (p, "true", 4)) {
+				js = create_json (R_JSON_BOOL, key, parent);
+				js->num.u_value = 1;
+				return p + 4;
+			}
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0; // error
+		case 'f':
+			if (!strncmp (p, "false", 5)) {
+				js = create_json (R_JSON_BOOL, key, parent);
+				js->num.u_value = 0;
+				return p + 5;
+			}
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0; // error
+		case 'n':
+			if (!strncmp (p, "null", 4)) {
+				create_json (R_JSON_NULL, key, parent);
+				return p + 4;
+			}
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0; // error
+		case '/': // comment
+			if (p[1] == '/') { // line comment
+				char *ps = p;
+				p = strchr (p + 2, '\n');
+				if (!p) {
+					R_JSON_REPORT_ERROR ("endless comment", ps);
+					return 0; // error
+				}
+				p++;
+			} else if (p[1] == '*') { // block comment
+				p = skip_block_comment (p + 2);
+				if (!p) {
+					return 0;
+				}
+			} else {
+				R_JSON_REPORT_ERROR ("unexpected chars", p);
+				return 0; // error
+			}
+			break;
+		default:
+			R_JSON_REPORT_ERROR ("unexpected chars", p);
+			return 0; // error
+		}
+	}
+}
+
+R_API RJson *r_json_parse(char *text) {
```

Comment:
Const char *text

Reply:
All of these must be non-const because the parser works in-place and modifies the string. If you want to parse a const char * you have to strdup it yourself before.

Reply:
Or use pascal style strings like js0n do

Reply:
What do pascal style strings have to do with this?

Reply:
instead of null terminate the strings specify the length